test/clonevolume_tester.py

```diff
@@ -0,0 +1,262 @@
+        mock_etcd.get_vol_byname.return_value = data.volume_flash_cache
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        # Make addVolumeToVolumeSet fail by throwing exception
+        mock_3parclient.addVolumeToVolumeSet.side_effect = [exceptions.HTTPNotFound('fake')]
+
+
+# CHAP enabled makes Offline copy flow to execute
+class TestCloneWithCHAP(CloneVolumeUnitTest):
+    def override_configuration(self, config):
+        config.hpe3par_iscsi_chap_enabled = True
+
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.getVolumeMetaData.assert_called()
+        mock_3parclient.createVolume.assert_called()
+        mock_3parclient.copyVolume.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+        mock_3parclient.getVolumeMetaData.return_value = {'value': True}
+        mock_3parclient.getTask.return_value = {'status': data.TASK_DONE}
+
+# TODO: This is already covered in other tests above
+class TestCloneWithoutCHAP(CloneVolumeUnitTest):
+    pass
+
+
+# Override WS-API-Version to have lower value
+class TestCloneUnsupportedDedupVersion(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume_dedup
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.getWsApiVersion.return_value = \
+            {'major': 1,
+             # Setting it to lower version that doesn't support dedup
+             'build': 20301215,
+             'minor': 6,
+             'revision': 0}
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+
+
+# TODO: Compression related TCs to be added later
+class TestCloneCompressedVolume(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
```

Comment:
it is missing 'compression' parameter in the Opts

Reply:
That's right! We in fact added this parameter during our last session and the code I think is available in my latest commit on branch unit_test_fw_optimization.